### PR TITLE
Sheltered Ring / Brachyura Earring / Effect Audit

### PR DESCRIPTION
**Pull request by neuromancerxi**
Adds MOD ENHANCES_PROT_SHELL_RCVD (977) to the following:
Sheltered Ring 10764
Brachyura Earring 11039

These items add the following bonus when receiving spell/effect:
Add (Tier of Spell x 2) DEF to Protect Spells/Effects
Add (Tier of Spell / 256) to Shell Spells/Effects

Effect Detail and confirmation items do not stack:

https://www.ffxiah.com/item/10764/sheltered-ring
https://www.ffxiah.com/item/11039/brachyura-earring

Updates Spells/Items to use the current BG Values:

https://www.bg-wiki.com/bg/Protect (and _II -> _V)
https://www.bg-wiki.com/bg/Protectra (and _II -> _V)
https://www.bg-wiki.com/bg/Shell (and _II -> _V)
https://www.bg-wiki.com/bg/Shellra (and _II -> _V)
```
Protect/ra  DEF            DEF
                           With MOD
       I -   20             22
      II -   50             54
     III -   90             96
      IV -  140            148
       V -  220            230

Shell/ra    MDT  as /256   MDT  as /256
                           With MOD
       I -  -11 (-27/256)  -11 (-28/256)
      II -  -16 (-42/256)  -17 (-44/256)
     III -  -22 (-56/256)  -23 (-59/256)
      IV -  -26 (-67/256)  -28 (-71/256)
       V -  -29 (-75/256)  -31 (-80/256)

Shell/ra    BG MDT-        Value as Decimal     Rounded Value /10000
I           27/256         .10546875            1055
II          42/256         .1640625             1641
III         56/256         .21875               2188
IV          67/256         .26171875            2617
V           75/256         .29296875            2930

Bonus        1/256         .00390625            0039
```

Script Adjustments:

Added checks on individual Spells/Item Scripts.
Adjusts power value for Shell/ra spells/items to reflect /10000 scale.
Check for MOD Value > 0 (prevents stacking).
Updated Protect Values in scripts to match BG Values.
Added tier values for spells/effects.
Updated regimes to include Tier V and apply bonus with MOD present.

**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] that I agree to LandSandBoat's [Limited Contributor License Agreement](https://github.com/LandSandBoat/server/blob/base/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I have **read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)**
- [x] that I've _**tested my code and things my code changed**_ since the last commit in the PR, and will test after any later commits

### CODE REVIEW

File: scripts/globals/items/coated_shield.lua [hunk 1]
```diff
@@ -13,7 +14,14 @@ item_object.onItemCheck = function(target)
...
 
 item_object.onItemUse = function(target)
-    if (target:addStatusEffect(xi.effect.SHELL, 9, 0, 1800)) then
+    local power = 27 -- power/256 handled below to pass final DMGMAGIC value
+    local tier = 1
+    local buff = 0
+    if target:getMod(xi.mod.ENHANCES_PROT_SHELL_RCVD) > 0 then
+        buff = 1 -- Adds the tier as a bonus to power before calculation
+    end
+    local power = utils.roundup((power + (buff * tier)) / 2.56) -- takes the result and converts it back to a usable DMGMAGIC value
```
Comment (Xaver-DaRed) [inline scripts/globals/items/coated_shield.lua]: You are re-defining the variable. This happens in most scripts, so I'm only gonna point it out here, but applies to all of them.

File: scripts/globals/spells/shell_ii.lua [hunk 2]
```diff
@@ -12,16 +13,27 @@ spell_object.onMagicCastingCheck = function(caster, target, spell)
 end
 
 spell_object.onSpellCast = function(caster, target, spell)
-    local power = 14
+    local power = 42 -- power/256 handled below before passing final DMGMAGIC value
+    local tier =2
```
Comment (Xaver-DaRed) [inline scripts/globals/spells/shell_ii.lua]: This line needs alignment. Trailing whitespace

File: scripts/globals/items/coated_shield.lua (hunk 1 above)
Comment (TeoTwawki) [reply scripts/globals/items/coated_shield.lua]: its also not necessary to use values based on a 256ths scale and then do conversions despite what wiki says. SE uses those fractional values to save some bit programmatically, we don't have to match that and can just use the intended % to start with.

**Comment by Xaver-DaRed** [comment]
All in all, I think this is great.
The only thing I would change is the roundup and the initial power values with it.

Setting the initial power as a % simplifies the operations and makes the roundup obsolete.

**Comment by neuromancerxi** [comment]
> All in all, I think this is great.
> The only thing I would change is the roundup and the initial power values with it.
> 
> Setting the initial power as a % simplifies the operations and makes the roundup obsolete.

I tried it initially without using a Round-up function. During testing, I found that without it, certain values fall off or end up incorrect altogether. Dropping round-up into utils seemed like the best way to call it across the 12 or so scripts that needed it.

Another way I considered tackling it was to convert all of the DT/PDT/MDT values into 256ths, but abandoned that as way too big an undertaking to accomplish this smaller task.

edit: Thanks for the feedback and review!

**Comment by TeoTwawki** [comment]
> Another way I considered tackling it was to convert all of the DT/PDT/MDT values into 256ths, but abandoned that as way too big an undertaking to accomplish this smaller task.

yeah don't do that. they are intentionally not base 256 anymore here. this has changed multiple times and they were at one point 256 based which resulted in many contributors setting incorrect values. They don't need to precisely match to the decimal point whatever the result of the fractions would have been and there is no need for any of this converting.

**Comment by Xaver-DaRed** [comment]
Magic, in general, is going to recieve a severe makeover "soon".
Until then, I think that any fall off in power can/should be ignored, in favor of simplicity.

**Comment by neuromancerxi** [comment]
> > Another way I considered tackling it was to convert all of the DT/PDT/MDT values into 256ths, but abandoned that as way too big an undertaking to accomplish this smaller task.
> 
> yeah don't do that. they are intentionally not base 256 anymore here. this has changed multiple times and they were at one point 256 based which resulted in many contributors setting incorrect values. They don't need to precisely match to the decimal point whatever the result of the fractions would have been and there is no need for any of this converting.

I marked resolved your earlier feedback before I saw it Teo, sorry about that. During initial testing, I found that since the calculations are based on 256ths, there are some nuanced differences in results if you don't properly math it out against that scale.

If we start with just the MDT values and apply the bonus without rounding up, the numbers start to fall outside of what they should look like (lower or higher in some cases). I'm trying to keep the formula resulting in consistently accurate results as they've been reported by the community. The way I came up with to do that was to math out on 256ths.

If you have a cleaner/saner mechanism you're recommending, I'm open to suggestions and will see if I can replicate the accurate results using it.

**Comment by TeoTwawki** [comment]
a byte is `0x00` to `0xFF` for 255. 256 possible values in there in total (including zero).

we use a percent. take your fractions and convert to a percent, you will see we had the right power level for almost everything already. 75/256 = 29% and we had 29 for shellra V's script. changing it to 256 and back again you still wind up at 29 if you are doing it correctly anyway (if you arrived at anything else, you've made a mistake).

if you have any provable case where you need finer precision for a non whole percent point, let me know and I will make it use more digits so you can account for it (look what we did to haste years ago in darkstar). I want contributors to never have to think about how the number is represented like this.

tl;dr just use the percent value please.

**Comment by neuromancerxi** [comment]
Thanks for taking the time to reach out Teo. I'll see about reworking this request to work on a percent instead of 256.

**Comment by neuromancerxi** [comment]
The mechanic of this ring/earring (at least as folks in the community have puzzled out) is to add `tier/256` to the existing shell MDT value also expressed as `value/256`. If we convert these values (1/256 -> 5/256) into percents, we end up with .4 % (.004) steps. Since the buff being applied is so small, my concern in using these converted values (instead of the /256 value) is that they would have no appreciable affect due to the simplification of the percent values.

Example of results when working with already converted values (ex 1):

```
-11 * 1.004 = -11.044 (no effective buff)
-16 * 1.008 = -16.128 (no effective buff)
-22 * 1.012 = -22.264 (no effective buff)
-26 * 1.016 = -26.416 (no effective buff)
-29 * 1.020 = -29.58  (round up to 30)
```

Example of calculating the base shell MDT as /256 values (ex 2):

```
Base             W/Buff
-11 (-27/256)    -11 (-28/256)
-16 (-42/256)    -17 (-44/256)
-22 (-56/256)    -23 (-59/256)
-26 (-67/256)    -28 (-71/256)
-29 (-75/256)    -31 (-80/256)
```

There is a loss of granularity due to the converted values we're working having already been rounded down or up:

```
/256:       DEC:    Expressed/Rounded Value:
(-27/256)   .1054   -11 (LSB currently uses 9)
(-42/256)   .1640   -16
(-56/256)   .2187   -22
(-67/256)   .2617   -26
(-75/256)   .2929   -29
```

Putting it all together, the buff gives expected results when working in 256ths:

```
MDT Value:      DEC:    Buff Value: DEC:    Final MDT Val:  DEC:
-11 (-27/256)   .1054   (-1/256)    .0039   -11 (-28/256)   0.1093 (Round Up to 11)
-16 (-42/256)   .1640   (-2/256)    .0078   -17 (-44/256)   0.1718 (17)
-22 (-56/256)   .2187   (-3/256)    .0117   -23 (-59/256)   0.2304 (23)
-26 (-67/256)   .2617   (-4/256)    .0156   -28 (-71/256)   0.2773 (Round up to 28)
-29 (-75/256)   .2929   (-5/256)    .0195   -31 (-80/256)   0.3124 (31)
```

I want to make sure I understand how best to proceed before reworking the PR. Am I understanding correctly that the preference of the project is to start with the already adjusted decimal values and apply the item buff (ex 1) instead of using the more finite/accurate calculation (ex 2)?

Please let me know, and I'll plan accordingly.

Thanks!

**Comment by TeoTwawki** [comment]
I understood you just fine, you apparently still don't understand me. so I'm going to just throw it up myself for you to look at. I will not be playing any more of the show me the same math again game

**Review by Xaver-DaRed — APPROVED** [review]
